### DISCUSSION TITLE
Migration From Flectra To Odoo 14

### DISCUSSION THREAD

**Discussion by it12uw** [OCA/odoo-community.org#103] [Q&A] [answered]
hello guys, i want to ask a question, Has anyone here ever tried to migrate from Flectra to Odoo 14?
if so, how to do it and what are the risks of the migration process?

if there is, I beg for your help, because I want to migrate from flectra to odoo 14.

Thankyou guys 
God Bless :)

**Comment by legalsylvain** [comment]
Which version Of Flectra ?

**Reply by it12uw** [reply]
flectra 1.0 to odoo14?

is that possible?

**Reply by chienandalu** [reply, marked as answer]
It shouldn't be impossible with [OpenUpgrade](https://github.com/oca/openUpgrade) as it's just a fork of Odoo 11.0... Anyway, as it's a fork, things could diverge from the original and you might need to do your own migration scripts adaptations. Here's an introduction to the OpenUpgrade process in case you don't know it: https://www.youtube.com/watch?v=NbQ78YKrCX4

**Reply by pedrobaeza** [reply]
I would switch first from Flectra 1.0 to Odoo 11, updating whole database, and if everything goes OK, then think in migrating to upper versions of Odoo.

**Reply by it12uw** [reply]
okay, thank you guys for give your answer.

i'll try it.

**Reply by it12uw** [reply]
hello sir, can you please tell me how to export flectra database from postgreesql and then update it.?

thankyou sir

**Reply by pedrobaeza** [reply]
That depends on how Flectra is deployed on your side. Basically will be a pg_dump + pg_restore + update of the addons, but I suggest to contact an specialist to do it.

**Reply by it12uw** [reply]
oke, thankyou sir,
i have another question sir i wish that you can help my problem,
i got this error everytime i try to download more than 1 data records  
raise ValueError("Expected singleton: %s" % self)
ValueError: <class 'ValueError'>: "Expected singleton: account.invoice(35590, 35589, 35588, 35587)" while evaluating
'action = records.download_efaktur()'

here is my code for download_efaktur

@api.multi
    def download_efaktur(self):
        """Collect the data and execute function _generate_efaktur."""
        for fakturs in self:
            for record in fakturs:
                if record.state == 'draft':
                    raise ValidationError(_('Could not download E-faktur in draft state'))

                if record.partner_id.l10n_id_pkp and not record.l10n_id_tax_number:
                    raise ValidationError(_('Connect ') + record.name + _(' with E-faktur to download this report'))

            self._generate_efaktur(',')
        return self.download_csv()

**Reply by HekkiMelody** [reply]
You probably cut the part of the error that would have helped pinpoint exactly where. (Also please fix your formatting)

But in general the `Expected singleton` error happens every time you try to do something on a recordset of multiple records when the operation is only allowed on a single record.

I can't immediately see that in the bit of code you shared, so the issue may be in `self._generate_efaktur` (should it be `fakturs._generate_efaktur`? Hard to tell when the indentation is so unclear). It's also unclear to me why there are two `for` loops nested inside one another without more context.

**Reply by it12uw** [reply]
i'll share the full error message and the code for this case.

THE ERROR MESSAGE

Error:
Flectra Server Error

Traceback (most recent call last):
  File "C:\Users\IT UW\odoo14\flectra\flectra\models.py", line 4434, in ensure_one
    _id, = self._ids
ValueError: too many values to unpack (expected 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\IT UW\odoo14\flectra\flectra\tools\safe_eval.py", line 350, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "c:\users\it uw\odoo14\mspcustomaddons\l10n_id_efaktur\models\account_move.py", line 138, in download_efaktur
    self._generate_efaktur(',')
  File "c:\users\it uw\odoo14\mspcustomaddons\l10n_id_efaktur\models\account_move.py", line 298, in _generate_efaktur
    output_head = self._generate_efaktur_invoice(delimiter)
  File "c:\users\it uw\odoo14\mspcustomaddons\l10n_id_efaktur\models\account_move.py", line 146, in _generate_efaktur_invoice
    company_id = self.company_id
  File "C:\Users\IT UW\odoo14\flectra\flectra\fields.py", line 947, in __get__
    record.ensure_one()
  File "C:\Users\IT UW\odoo14\flectra\flectra\models.py", line 4437, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: account.invoice(35590, 35589, 35588, 35587)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 658, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 314, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "C:\Users\IT UW\odoo14\flectra\flectra\tools\pycompat.py", line 87, in reraise
    raise value
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 700, in dispatch
    result = self._call_function(**self.params)
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 346, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "C:\Users\IT UW\odoo14\flectra\flectra\service\model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 339, in checked_call
    result = self.endpoint(*a, **kw)
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 944, in __call__
    return self.method(*args, **kw)
  File "C:\Users\IT UW\odoo14\flectra\flectra\http.py", line 522, in response_wrap
    response = f(*args, **kw)
  File "c:\users\it uw\odoo14\flectra\addons\web\controllers\main.py", line 1387, in run
    result = request.env['ir.actions.server'].browse([action_id]).run()
  File "c:\users\it uw\odoo14\flectra\flectra\addons\base\ir\ir_actions.py", line 558, in run
    res = func(action, eval_context=eval_context)
  File "c:\users\it uw\odoo14\flectra\flectra\addons\base\ir\ir_actions.py", line 434, in run_action_code_multi
    safe_eval(action.sudo().code.strip(), eval_context, mode="exec", nocopy=True)  # nocopy allows to return 'action'
  File "C:\Users\IT UW\odoo14\flectra\flectra\tools\safe_eval.py", line 373, in safe_eval
    pycompat.reraise(ValueError, ValueError('%s: "%s" while evaluating\n%r' % (ustr(type(e)), ustr(e), expr)), exc_info[2])
  File "C:\Users\IT UW\odoo14\flectra\flectra\tools\pycompat.py", line 86, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\IT UW\odoo14\flectra\flectra\tools\safe_eval.py", line 350, in safe_eval
    return unsafe_eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "c:\users\it uw\odoo14\mspcustomaddons\l10n_id_efaktur\models\account_move.py", line 138, in download_efaktur
    self._generate_efaktur(',')
  File "c:\users\it uw\odoo14\mspcustomaddons\l10n_id_efaktur\models\account_move.py", line 298, in _generate_efaktur
    output_head = self._generate_efaktur_invoice(delimiter)
  File "c:\users\it uw\odoo14\mspcustomaddons\l10n_id_efaktur\models\account_move.py", line 146, in _generate_efaktur_invoice
    company_id = self.company_id
  File "C:\Users\IT UW\odoo14\flectra\flectra\fields.py", line 947, in __get__
    record.ensure_one()
  File "C:\Users\IT UW\odoo14\flectra\flectra\models.py", line 4437, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: <class 'ValueError'>: "Expected singleton: account.invoice(35590, 35589, 35588, 35587)" while evaluating
'action = records.download_efaktur()'

THE CODE 
``
# -*- encoding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.

import base64
import re
from flectra import api, fields, models, _
from flectra.exceptions import UserError, ValidationError
from flectra.tools import float_round
from datetime import datetime, timedelta

FK_HEAD_LIST = ['FK', 'KD_JENIS_TRANSAKSI', 'FG_PENGGANTI', 'NOMOR_FAKTUR', 'MASA_PAJAK', 'TAHUN_PAJAK', 'TANGGAL_FAKTUR', 'NPWP', 'NAMA', 'ALAMAT_LENGKAP', 'JUMLAH_DPP', 'JUMLAH_PPN', 'JUMLAH_PPNBM', 'ID_KETERANGAN_TAMBAHAN', 'FG_UANG_MUKA', 'UANG_MUKA_DPP', 'UANG_MUKA_PPN', 'UANG_MUKA_PPNBM', 'REFERENSI', 'KODE_DOKUMEN_PENDUKUNG']

LT_HEAD_LIST = ['LT', 'NPWP', 'NAMA', 'JALAN', 'BLOK', 'NOMOR', 'RT', 'RW', 'KECAMATAN', 'KELURAHAN', 'KABUPATEN', 'PROPINSI', 'KODE_POS', 'NOMOR_TELEPON']

OF_HEAD_LIST = ['OF', 'KODE_OBJEK', 'NAMA', 'HARGA_SATUAN', 'JUMLAH_BARANG', 'HARGA_TOTAL', 'DISKON', 'DPP', 'PPN', 'TARIF_PPNBM', 'PPNBM']

def _csv_row(data, delimiter=',', quote='"'):
    return quote + (quote + delimiter + quote).join([str(x).replace(quote, '\\' + quote) for x in data]) + quote + '\n'

class AccountMove(models.Model):
    _inherit = "account.invoice"

    country_code = fields.Char(related='company_id.country_id.code', string='Country Code')
    l10n_id_tax_number = fields.Char(string="Tax Number", copy=False)
    l10n_id_replace_invoice_id = fields.Many2one('account.invoice', string="Replace Invoice",  domain="['|', '&', '&', ('state', '=', 'posted'), ('partner_id', '=', partner_id), ('reversal_move_id', '!=', False), ('state', '=', 'cancel')]", copy=False)
    l10n_id_attachment_id = fields.Many2one('ir.attachment', readonly=True, copy=False)
    l10n_id_csv_created = fields.Boolean('CSV Created', compute='_compute_csv_created', copy=False)
    l10n_id_kode_transaksi = fields.Selection([
            ('01', '01 Kepada Pihak yang Bukan Pemungut PPN (Customer Biasa)'),
            ('02', '02 Kepada Pemungut Bendaharawan (Dinas Kepemerintahan)'),
            ('03', '03 Kepada Pemungut Selain Bendaharawan (BUMN)'),
            ('04', '04 DPP Nilai Lain (PPN 1%)'),
            ('05', '05 Besaran Tertentu'),
            ('06', '06 Penyerahan Lainnya (Turis Asing)'),
            ('07', '07 Penyerahan yang PPN-nya Tidak Dipungut (Kawasan Ekonomi Khusus/ Batam)'),
            ('08', '08 Penyerahan yang PPN-nya Dibebaskan (Impor Barang Tertentu)'),
            ('09', '09 Penyerahan Aktiva ( Pasal 16D UU PPN )'),
        ], string='Kode Transaksi', help='Dua digit pertama nomor pajak',
        readonly=True, states={'draft': [('readonly', False)]}, copy=False)
    l10n_id_need_kode_transaksi = fields.Boolean(compute='_compute_need_kode_transaksi')
    tax_vendor_number = fields.Char(string="Tax Vendor")

    @api.onchange('partner_id')
    def _onchange_partner_id(self):
        self.l10n_id_kode_transaksi = self.partner_id.l10n_id_kode_transaksi
        return super(AccountMove, self)._onchange_partner_id()

    @api.onchange('l10n_id_tax_number')
    def _onchange_l10n_id_tax_number(self):
        for record in self:
            if record.l10n_id_tax_number and record.type not in self.get_purchase_types():
                raise UserError(_("You can only change the number manually for a Vendor Bills and Credit Notes"))

    @api.depends('l10n_id_attachment_id')
    def _compute_csv_created(self):
        for record in self:
            record.l10n_id_csv_created = bool(record.l10n_id_attachment_id)

    @api.depends('partner_id')
    def _compute_need_kode_transaksi(self):
        for move in self:
            move.l10n_id_need_kode_transaksi = move.partner_id.l10n_id_pkp and not move.l10n_id_tax_number and move.type == 'out_invoice' and move.country_code == 'ID'

    # @api.constrains('l10n_id_kode_transaksi', 'line_ids')
    # def _constraint_kode_ppn(self):
    #     ppn_tag = self.env.ref('l10n_id.ppn_tag')
    #     for move in self.filtered(lambda m: m.l10n_id_kode_transaksi != '08'):
    #         if any(ppn_tag.id in line.tag_ids.ids for line in move.line_ids if line.exclude_from_invoice_tab is False and not line.display_type) \
    #                 and any(ppn_tag.id not in line.tag_ids.ids for line in move.line_ids if line.exclude_from_invoice_tab is False and not line.display_type):
    #             raise UserError(_('Cannot mix VAT subject and Non-VAT subject items in the same invoice with this kode transaksi.'))
    #     for move in self.filtered(lambda m: m.l10n_id_kode_transaksi == '08'):
    #         if any(ppn_tag.id in line.tag_ids.ids for line in move.line_ids if line.exclude_from_invoice_tab is False and not line.display_type):
    #             raise UserError('Kode transaksi 08 is only for non VAT subject items.')

    @api.constrains('l10n_id_tax_number')
    def _constrains_l10n_id_tax_number(self):
        for record in self.filtered('l10n_id_tax_number'):
            if record.l10n_id_tax_number != re.sub(r'\D', '', record.l10n_id_tax_number):
                record.l10n_id_tax_number = re.sub(r'\D', '', record.l10n_id_tax_number)
            if len(record.l10n_id_tax_number) != 16:
                raise UserError(_('A tax number should have 16 digits'))
            elif record.l10n_id_tax_number[:2] not in dict(self._fields['l10n_id_kode_transaksi'].selection).keys():
                raise UserError(_('A tax number must begin by a valid Kode Transaksi'))
            elif record.l10n_id_tax_number[2] not in ('0', '1'):
                raise UserError(_('The third digit of a tax number must be 0 or 1'))

    def invoice_validate(self):
        """Set E-Faktur number after validation."""
        for move in self:
            if move.l10n_id_need_kode_transaksi:
                if not move.l10n_id_kode_transaksi:
                    raise ValidationError(_('You need to put a Kode Transaksi for this partner.'))
                if move.l10n_id_replace_invoice_id.l10n_id_tax_number:
                    if not move.l10n_id_replace_invoice_id.l10n_id_attachment_id:
                        raise ValidationError(_('Replacement invoice only for invoices on which the e-Faktur is generated. '))
                    rep_efaktur_str = move.l10n_id_replace_invoice_id.l10n_id_tax_number
                    move.l10n_id_tax_number = '%s1%s' % (move.l10n_id_kode_transaksi, rep_efaktur_str[3:])
                else:
                    efaktur = self.env['l10n_id_efaktur.efaktur.range'].pop_number(move.company_id.id)
                    if not efaktur:
                        raise ValidationError(_('There is no Efaktur number available.  Please configure the range you get from the government in the e-Faktur menu. '))
                    move.l10n_id_tax_number = '%s0%013d' % (str(move.l10n_id_kode_transaksi), efaktur)
        return super(AccountMove, self).invoice_validate()

    def reset_efaktur(self):
        """Reset E-Faktur, so it can be use for other invoice."""
        for move in self:
            if move.l10n_id_attachment_id:
                raise UserError(_('You have already generated the tax report for this document: %s') % move.name)
            self.env['l10n_id_efaktur.efaktur.range'].push_number(move.company_id.id, move.l10n_id_tax_number[3:])
            move.message_post(
                body='e-Faktur Reset: %s ' % (move.l10n_id_tax_number),
                subject="Reset Efaktur")
            move.l10n_id_tax_number = False
        return True

    def download_csv(self):
        action = {
            'type': 'ir.actions.act_url',
            'url': "web/content/?model=ir.attachment&id=" + str(self.l10n_id_attachment_id.id) + "&filename_field=name&field=datas&download=true&name=" + self.l10n_id_attachment_id.name,
            'target': 'self'
        }
        return action

    @api.multi
    def download_efaktur(self):
        """Collect the data and execute function _generate_efaktur."""
        for fakturs in self:
            for record in fakturs:
                if record.state == 'draft':
                    raise ValidationError(_('Could not download E-faktur in draft state'))

                if record.partner_id.l10n_id_pkp and not record.l10n_id_tax_number:
                    raise ValidationError(_('Connect ') + record.name + _(' with E-faktur to download this report'))

            self._generate_efaktur(',')
        return self.download_csv()

    @api.multi
    def _generate_efaktur_invoice(self, delimiter):
        """Generate E-Faktur for customer invoice."""
        # Invoice of Customer
        for invoice in self:
            company_id = self.company_id
            dp_product_id = self.env['ir.config_parameter'].sudo().get_param('sale.default_deposit_product_id')

            output_head = '%s%s%s' % (
                _csv_row(FK_HEAD_LIST, delimiter),
                _csv_row(LT_HEAD_LIST, delimiter),
                _csv_row(OF_HEAD_LIST, delimiter),
            )
            for move in invoice.filtered(lambda m: m.state not in ('draft','cancel')):
                eTax = move._prepare_etax()

                nik = str(move.partner_id.l10n_id_nik) if not move.partner_id.vat else ''

                if move.l10n_id_replace_invoice_id:
                    number_ref = str(move.l10n_id_replace_invoice_id.move_name) + " replaced by " + str(move.move_name) + " " + nik
                else:
                    number_ref = str(move.move_name) + " " + nik

                street = ', '.join([x for x in (move.partner_id.street, move.partner_id.street2) if x])

                invoice_npwp = '000000000000000'
                if move.partner_id.vat and len(move.partner_id.vat) >= 12:
                    invoice_npwp = move.partner_id.vat
                elif (not move.partner_id.vat or len(move.partner_id.vat) < 12) and move.partner_id.l10n_id_nik:
                    invoice_npwp = move.partner_id.l10n_id_nik
                invoice_npwp = invoice_npwp.replace('.', '').replace('-', '')
                date_invoice = datetime.strptime(move.date_invoice, '%Y-%m-%d')
                # Here all fields or columns based on eTax Invoice Third Party
                eTax['KD_JENIS_TRANSAKSI'] = move.l10n_id_tax_number[0:2] or 0
                eTax['FG_PENGGANTI'] = move.l10n_id_tax_number[2:3] or 0
                eTax['NOMOR_FAKTUR'] = move.l10n_id_tax_number[3:] or 0
                eTax['MASA_PAJAK'] = date_invoice.month
                eTax['TAHUN_PAJAK'] = date_invoice.year
                eTax['TANGGAL_FAKTUR'] = '{0}/{1}/{2}'.format(date_invoice.day, date_invoice.month, date_invoice.year)
                eTax['NPWP'] = invoice_npwp
                eTax['NAMA'] = move.partner_id.name if eTax['NPWP'] == '000000000000000' else move.partner_id.l10n_id_tax_name or move.partner_id.name
                eTax['ALAMAT_LENGKAP'] = move.partner_id.contact_address.replace('\n', '') if eTax['NPWP'] == '000000000000000' else move.partner_id.l10n_id_tax_address or street
                eTax['JUMLAH_DPP'] = int(float_round(move.amount_untaxed, 0)) # currency rounded to the unit
                eTax['JUMLAH_PPN'] = int(float_round(move.amount_tax, 0))
                eTax['ID_KETERANGAN_TAMBAHAN'] = '1' if move.l10n_id_kode_transaksi == '07' else ''
                eTax['REFERENSI'] = number_ref
                eTax['KODE_DOKUMEN_PENDUKUNG'] = '0'

                lines = move.invoice_line_ids.filtered(lambda x: x.product_id.id == int(dp_product_id) and x.price_unit < 0 and not x.display_type)
                eTax['FG_UANG_MUKA'] = 0
                eTax['UANG_MUKA_DPP'] = int(abs(sum(lines.mapped(lambda l: float_round(l.price_subtotal, 0)))))
                eTax['UANG_MUKA_PPN'] = int(abs(sum(lines.mapped(lambda l: float_round(l.price_total - l.price_subtotal, 0)))))

                company_npwp = company_id.partner_id.vat or '000000000000000'

                fk_values_list = ['FK'] + [eTax[f] for f in FK_HEAD_LIST[1:]]
                eTax['JALAN'] = company_id.partner_id.l10n_id_tax_address or company_id.partner_id.street
                eTax['NOMOR_TELEPON'] = company_id.phone or ''

                lt_values_list = ['FAPR', company_npwp, company_id.name] + [eTax[f] for f in LT_HEAD_LIST[3:]]

                # HOW TO ADD 2 line to 1 line for free product
                free, sales = [], []

                for line in move.invoice_line_ids:
                    # *invoice_line_unit_price is price unit use for harga_satuan's column
                    # *invoice_line_quantity is quantity use for jumlah_barang's column
                    # *invoice_line_total_price is bruto price use for harga_total's column
                    # *invoice_line_discount_m2m is discount price use for diskon's column
                    # *line.price_subtotal is subtotal price use for dpp's column
                    # *tax_line or free_tax_line is tax price use for ppn's column
                    free_tax_line = tax_line = bruto_total = total_discount = 0.0

                    for tax in line.invoice_line_tax_ids:
                        if tax.amount > 0:
                            tax_line += line.price_subtotal * (tax.amount / 100.0)

                    discount = 1 - (line.discount / 100)
                    # guarantees price to be tax-excluded
                    invoice_line_total_price = line.price_subtotal / discount if discount else 0
                    invoice_line_unit_price = invoice_line_total_price / line.quantity if line.quantity else 0

                    line_dict = {
                        'KODE_OBJEK': line.product_id.default_code or '',
                        'NAMA': line.product_id.name or '',
                        'HARGA_SATUAN': int(float_round(invoice_line_unit_price, 0)),
                        'JUMLAH_BARANG': line.quantity,
                        'HARGA_TOTAL': int(float_round(invoice_line_total_price, 0)),
                        'DPP': int(float_round(line.price_subtotal, 0)),
                        'product_id': line.product_id.id,
                    }

                    if line.price_subtotal < 0:
                        for tax in line.invoice_line_tax_ids:
                            free_tax_line += (line.price_subtotal * (tax.amount / 100.0)) * -1.0

                        line_dict.update({
                            'DISKON': int(float_round(invoice_line_total_price - line.price_subtotal, 0)),
                            'PPN': int(float_round(free_tax_line, 0)),
                        })
                        free.append(line_dict)
                    elif line.price_subtotal != 0.0:
                        invoice_line_discount_m2m = invoice_line_total_price - line.price_subtotal

                        line_dict.update({
                            'DISKON': int(float_round(invoice_line_discount_m2m, 0)),
                            'PPN': int(float_round(tax_line, 0)),
                        })
                        sales.append(line_dict)

                sub_total_before_adjustment = sub_total_ppn_before_adjustment = 0.0

                # We are finding the product that has affected
                # by free product to adjustment the calculation
                # of discount and subtotal.
                # - the price total of free product will be
                # included as a discount to related of product.
                for sale in sales:
                    for f in free:
                        if f['product_id'] == sale['product_id']:
                            sale['DISKON'] = sale['DISKON'] - f['DISKON'] + f['PPN']
                            sale['DPP'] = sale['DPP'] + f['DPP']

                            tax_line = 0

                            for tax in line.invoice_line_tax_ids:
                                if tax.amount > 0:
                                    tax_line += sale['DPP'] * (tax.amount / 100.0)

                            sale['PPN'] = int(float_round(tax_line, 0))

                            free.remove(f)

                    sub_total_before_adjustment += sale['DPP']
                    sub_total_ppn_before_adjustment += sale['PPN']
                    bruto_total += sale['DISKON']
                    total_discount += float_round(sale['DISKON'], 2)

                output_head += _csv_row(fk_values_list, delimiter)
                output_head += _csv_row(lt_values_list, delimiter)
                for sale in sales:
                    of_values_list = ['OF'] + [str(sale[f]) for f in OF_HEAD_LIST[1:-2]] + ['0', '0']
                    output_head += _csv_row(of_values_list, delimiter)

        return output_head

    def _prepare_etax(self):
        # These values are never set
        return {'JUMLAH_PPNBM': 0, 'UANG_MUKA_PPNBM': 0, 'BLOK': '', 'NOMOR': '', 'RT': '', 'RW': '', 'KECAMATAN': '', 'KELURAHAN': '', 'KABUPATEN': '', 'PROPINSI': '', 'KODE_POS': '', 'JUMLAH_BARANG': 0, 'TARIF_PPNBM': 0, 'PPNBM': 0}

    @api.multi
    def _generate_efaktur(self, delimiter):
        if self.filtered(lambda x: not x.l10n_id_kode_transaksi):
            raise UserError(_('Some documents don\'t have a transaction code'))
        if self.filtered(lambda x: x.type != 'out_invoice'):
            raise UserError(_('Some documents are not Customer Invoices'))

        output_head = self._generate_efaktur_invoice(delimiter)
        my_utf8 = output_head.encode("utf-8")
        out = base64.b64encode(my_utf8)

        attachment = self.env['ir.attachment'].create({
            'datas': out,
            'name': 'efaktur_%s.csv' % (fields.Datetime.to_string(datetime.now()).replace(" ", "_")),
            'type': 'binary',
        })

        for record in self:
            record.message_post(attachment_ids=[attachment.id])
        self.l10n_id_attachment_id = attachment.id
        return {
            'type': 'ir.actions.client',
            'tag': 'reload',
        }

``

**Reply by HekkiMelody** [reply]
So I was correct, the issue is here `self._generate_efaktur(',')` this method can only be called on a single record, not on a recordset.

You probably want to call `fakturs._generate_efaktur()` or `record._generate_efaktur()`

**Reply by it12uw** [reply]
any idea to fix the `self._generate_efaktur` so it can be call on a recordset.?

**Reply by HekkiMelody** [reply]
Going to echo the others in this thread, that if you need dedicated step-by-step assistance, you contact a specialist and/or ask for help with the Flectra project, rather than with Odoo, sorry. Good luck with your migration!

**Comment by legalsylvain** [comment]
Well, your problem is related to flectra, not odoo. please open an issue on the dedicated project.
